core/common/src/main/java/alluxio/grpc/GrpcConnectionPool.java

```diff
@@ -0,0 +1,378 @@
+        PropertyKey.Template.USER_NETWORK_NETTY_CHANNEL
+            .format(key.getNetworkGroup().getPropertyCode()),
+        conf);
+    EventLoopGroup eventLoopGroup = acquireNetworkEventLoop(key, conf);
+
+    // Update the builder.
+    channelBuilder.keepAliveTime(keepAliveTimeMs, TimeUnit.MILLISECONDS);
+    channelBuilder.keepAliveTimeout(keepAliveTimeoutMs, TimeUnit.MILLISECONDS);
+    channelBuilder.maxInboundMetadataSize((int) inboundMessageSizeBytes);
+    channelBuilder.flowControlWindow((int) flowControlWindow);
+    channelBuilder.channelType(channelType);
+    channelBuilder.eventLoopGroup(eventLoopGroup);
+    // Use plaintext
+    channelBuilder.usePlaintext();
+  }
+
+  /**
+   * Returns {@code true} if given managed channel is ready.
+   */
+  private boolean waitForConnectionReady(ManagedChannel managedChannel, AlluxioConfiguration conf) {
+    long healthCheckTimeoutMs = conf.getMs(PropertyKey.NETWORK_CONNECTION_HEALTH_CHECK_TIMEOUT);
+    try {
+      Boolean res = CommonUtils.waitForResult("channel to be ready", () -> {
+        ConnectivityState currentState = managedChannel.getState(true);
+        switch (currentState) {
+          case READY:
+            return true;
+          case TRANSIENT_FAILURE:
+          case SHUTDOWN:
+            return false;
+          case IDLE:
+          case CONNECTING:
+            return null;
+          default:
+            return null;
+        }
+      }, (b) -> b != null, WaitForOptions.defaults().setTimeoutMs((int) healthCheckTimeoutMs));
+      return res;
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      return false;
+    } catch (TimeoutException e) {
+      return false;
+    }
+  }
+
+  /**
+   * Tries to gracefully shut down the managed channel. If falls back to forceful shutdown if
+   * graceful shutdown times out.
+   */
+  private void shutdownManagedChannel(ManagedChannel managedChannel, AlluxioConfiguration conf) {
+    // Close the gRPC managed-channel if not shut down already.
+    if (!managedChannel.isShutdown()) {
+      long gracefulTimeoutMs = conf.getMs(PropertyKey.NETWORK_CONNECTION_SHUTDOWN_GRACEFUL_TIMEOUT);
+      managedChannel.shutdown();
+      try {
+        if (!managedChannel.awaitTermination(gracefulTimeoutMs, TimeUnit.MILLISECONDS)) {
+          LOG.warn("Timed out gracefully shutting down connection: {}. ", managedChannel);
+        }
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        // Allow thread to exit.
+      }
+    }
+    // Forceful shut down if still not terminated.
+    if (!managedChannel.isTerminated()) {
+      long timeoutMs = conf.getMs(PropertyKey.NETWORK_CONNECTION_SHUTDOWN_TIMEOUT);
+
+      managedChannel.shutdownNow();
+      try {
+        if (!managedChannel.awaitTermination(timeoutMs, TimeUnit.MILLISECONDS)) {
+          LOG.warn("Timed out forcefully shutting down connection: {}. ", managedChannel);
```

Comment:
Is there any additional action we can take here?

Reply:
Not really.